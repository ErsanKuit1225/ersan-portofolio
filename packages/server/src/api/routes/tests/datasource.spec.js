const { 
  supertest,
} = require("./utilities")
let TestConfig = require("./utilities/TestConfiguration")
let { basicDatasource } = require("./utilities/structures")
let { checkBuilderEndpoint } = require("./utilities/TestFunctions")

describe("/datasources", () => {
  let request
  let server
  let config

  beforeAll(async () => {
    ({ request, server } = await supertest())
    config = new TestConfig(request)
  });

  afterAll(() => {
    server.close()
  })

  beforeEach(async () => {
    await config.init()
  });

  describe("create", () => {
    it("should create a new datasource", async () => {
      const res = await request
        .post(`/api/datasources`)
        .send(basicDatasource())
        .set(config.defaultHeaders())
        .expect('Content-Type', /json/)
        .expect(200)

        expect(res.res.statusMessage).toEqual("Datasource saved successfully.");            
        expect(res.body.name).toEqual("Test");
      })
    });

  describe("fetch", () => {
    let datasource

    beforeEach(async () => {
      datasource = await config.createDatasource()
    });

    afterEach(() => {
      delete datasource._rev
    });

    it("returns all the datasources from the server", async () => {
      const res = await request
        .get(`/api/datasources`)
        .set(config.defaultHeaders())
        .expect('Content-Type', /json/)
        .expect(200)

        const datasources = res.body
        expect(datasources).toEqual([
          {
            "_id": datasources[0]._id,
            "_rev": datasources[0]._rev,
            ...basicDatasource()
          }
        ]);            
    })

    it("should apply authorization to endpoint", async () => {
        await checkBuilderEndpoint({
          config,
          method: "GET",
          url: `/api/datasources`,
        })
      })
    });

  describe("destroy", () => {
    let datasource

    beforeEach(async () => {
      datasource = await config.createDatasource()
    });

    afterEach(() => {
      delete datasource._rev
    });

    it("deletes queries for the datasource after deletion and returns a success message", async () => {
      await config.createQuery()

      await request
        .delete(`/api/datasources/${datasource._id}/${datasource._rev}`)
        .set(config.defaultHeaders())
        .expect(200)

      const res = await request
        .get(`/api/datasources`)
        .set(config.defaultHeaders())
        .expect('Content-Type', /json/)
        .expect(200)
      
        expect(res.body).toEqual([])
    })

    it("should apply authorization to endpoint", async () => {
      await checkBuilderEndpoint({
        config,
        method: "DELETE",
        url: `/api/datasources/${datasource._id}/${datasource._rev}`,
      })
    })

  });
});
