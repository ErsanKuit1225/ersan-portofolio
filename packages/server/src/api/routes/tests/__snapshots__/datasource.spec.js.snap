// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/datasources fetch returns all the datasources from the server 1`] = `
Array [
  Object {
    "_id": "bb_internal",
    "config": Object {},
    "entities": Array [
      Object {
        "_id": "ta_users",
        "_rev": "1-039883a06c1f9cb3945731d79838181e",
        "name": "Users",
        "primaryDisplay": "email",
        "schema": Object {
          "email": Object {
            "constraints": Object {
              "email": true,
              "length": Object {
                "maximum": "",
              },
              "presence": true,
              "type": "string",
            },
            "fieldName": "email",
            "name": "email",
            "type": "string",
          },
          "firstName": Object {
            "constraints": Object {
              "presence": false,
              "type": "string",
            },
            "fieldName": "firstName",
            "name": "firstName",
            "type": "string",
          },
          "lastName": Object {
            "constraints": Object {
              "presence": false,
              "type": "string",
            },
            "fieldName": "lastName",
            "name": "lastName",
            "type": "string",
          },
          "roleId": Object {
            "constraints": Object {
              "inclusion": Array [
                "ADMIN",
                "POWER",
                "BASIC",
                "PUBLIC",
              ],
              "presence": false,
              "type": "string",
            },
            "fieldName": "roleId",
            "name": "roleId",
            "type": "options",
          },
          "status": Object {
            "constraints": Object {
              "inclusion": Array [
                "active",
                "inactive",
              ],
              "presence": false,
              "type": "string",
            },
            "fieldName": "status",
            "name": "status",
            "type": "options",
          },
        },
        "type": "table",
        "views": Object {},
      },
    ],
    "name": "Budibase DB",
    "source": "BUDIBASE",
    "type": "budibase",
  },
  Object {
    "_id": "datasource_f8f81b1f0893478580b863fe96f1f3da",
    "_rev": "1-1df90f81a2294ba7349f690f4a6df092",
    "config": Object {},
    "name": "Test",
    "source": "POSTGRES",
    "type": "datasource",
  },
]
`;
